### PR TITLE
Improve Dependency Resolution for Workflow Optimization

### LINKED ISSUE #2235: Improve Dependency Resolution for Workflow Optimization
### Title:
Improve Dependency Resolution for Workflow Optimization

#### Summary:
The current implementation of our GitHub Action, as defined in [sanguine/contrib/git-changes-action](https://github.com/synapsecns/sanguine/tree/master/contrib/git-changes-action), uses `go.mod` to determine dependencies for triggering workflows. This broad approach can lead to unnecessary workflow runs, as changes in dependencies that do not directly affect the module in question still trigger the workflow. We propose enhancing our dependency resolution mechanism to be more granular, focusing on package-level dependencies rather than module-level dependencies defined in `go.mod`.

#### Problem:
Currently, if any dependency within a module changes, our workflow is triggered, regardless of whether the change directly impacts the packages our module uses. This leads to inefficient use of resources, as workflows run for changes that are irrelevant to the module's operation. For example, if our module `A` directly depends on package `A` from module `B`, we want to trigger the workflow only if package `A` (or any package within `A`'s dependency tree) changes. Changes in module `B` that do not affect package `A` should not trigger our module `A`'s workflow.

#### Proposed Solution:
Implement a more refined dependency tracking system that:
1. Identifies direct package-level dependencies for each module within our project.
2. Triggers workflows only when there are changes in the directly dependent packages or any package within their dependency tree.
3. Excludes broader module-level changes from triggering workflows when the changes do not affect the directly used packages.

This approach aims to optimize our CI/CD process by reducing unnecessary workflow runs, thereby saving computational resources and developer time.

#### Expected Benefits:
- **Efficiency:** Reduces unnecessary CI/CD runs, focusing only on relevant changes.
- **Resource Optimization:** Better use of computational resources, decreasing overall run times and potential costs.
- **Developer Productivity:** Developers can focus on changes that directly impact their work, reducing noise from unrelated workflow runs.

#### Request:
We seek input from the community and the project's maintainers on the feasibility of this approach and any suggestions for implementation strategies. Additionally, we're looking for contributors interested in developing this enhancement.

### PR DESCRIPTION
**Pull request by oliverqx**
/claim #2235

**Description**
This PR implements module package-based dependency resolution as requested by issue #2235. 

**Additional context**
Linked issue explains it quite well.

**Metadata**
- Fixes #2235

## Summary by CodeRabbit

- **New Features**
  - Enhanced the `Git Changes Action` to export a variable with changed Go modules in pull requests, introducing `Module to module` and `Module to package` dependency resolution methods.
  - Customization options include `levelOfDependencyResolution` for specifying resolution method and examples for testing based on modified modules and dependencies.
- **Refactor**
  - Updated function calls in the main function to include an additional argument for specifying the type of dependency.
- **Documentation**
  - README.md now provides details on new features, dependency resolution methods, and examples for testing with different resolution levels.

### CODE REVIEW

**Comment by trajan0x** [comment]
@oliverqx btw working on some changes to let these wokflows run correctly

**Comment by oliverqx** [comment]
@trajan0x I created this [repository](https://github.com/oliverqx/sanguine-git-changes-action) (which is a synapse/sanguine clone without all the git history and without unrelated github workflows) where I've been running the github action. Would be nice to have the action tested there to avoid polluting this repository. I'd like to open some PRs as test cases so I've been reading the repo and I'm struggling to find the following case:

```
ModuleA/PackageA -> ModuleB/PackageB -> ModuleC/PackageC
```
> where -> represents  'depends on'

Conditions:
- ModuleA/PackageA should not import ModuleC/PackageC directly

Once a case with those conditions is found, the plan is to open two PRs:
1. Change ModuleC/PackageC and have the action flag ModuleA and ModuleB as modified.
2. Change ModuleC/PackageA and prove action does not flag ModuleA and ModuleB as modified.

Do you by any change know of such a relationship within the codebase?

File: contrib/git-changes-action/detector/detector.go [hunk 1]
```diff
@@ -41,28 +41,62 @@
...
+          
+          if changed {
+             break
+          }
+        }
+      }
+
+      modules[module.Path] = changed
+    }
+  }
```
[the remark this answers is not in the document]
Comment (oliverqx) [reply contrib/git-changes-action/detector/detector.go]: will resolve conversation as this code is not the most recent commit

File: contrib/git-changes-action/detector/detector.go [hunk 2]
```diff
@@ -19,16 +19,43 @@ import (
...
+
+      if changed == true {
+        break
+      }
+    }
+  }
+
+  packages[packageName] = changed
+  return changed
+}
```
[the remark this answers is not in the document]
Comment (oliverqx) [reply contrib/git-changes-action/detector/detector.go]: will resolve conversation as this code is not the most recent commit

File: contrib/git-changes-action/detector/detector.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (oliverqx) [reply contrib/git-changes-action/detector/detector.go]: can you review the most recent changes?

File: contrib/git-changes-action/detector/detector.go [hunk 3]
```diff
@@ -41,28 +41,69 @@
...
+          // the module can be flagged as changed
+          if changed {
+             break
+          }
+        }
+      }
+
+      modules[module.Path] = changed
+    }
+  }
```
[the remark this answers is not in the document]
Comment (oliverqx) [reply contrib/git-changes-action/detector/detector.go]: took the advice, will resolve conversation

**Comment by trajan0x** [comment]
sorry @oliverqx seeing this isnt' a draft anymore. Is this ready for revie?

**Comment by oliverqx** [comment]
@trajan0x It is..  about the structure of my solution. I added a parameter to the detector function to maintain module level dependency resolution as an option. This creates two logic branches to all subsequent functions, and its a bit strange. I was thinking I could create two subpackages in detector:

1. moduledetector: will contain function, utilities and tests to detect changed modules on module level deps
2. packagedetector: will contain function, utilities and tests, to detect changed modules on package level deps
3. detector: will contain git related functions (getchangetree, gethead, etc)

This fixes a lot of the lint issues and, I think, its a better dx.  I've already done it locally but since its quite a structural change, I wanted to ask if thats a good route to take before I push the changes.

**Comment by trajan0x** [comment]
@oliverqx think this makes a lot of sense

**Comment by oliverqx** [comment]
@trajan0x  just pushed the changes, i think its ready to be reviewed

**Comment by trajan0x** [comment]
fyi, available to try no w/ docker tag ghcr.io/synapsecns/sanguine/git-changes-action:07a4696bbbafbba2d3ad7ee828b4c8adb21b4d20

**Comment by oliverqx** [comment]
I created this [repo](https://github.com/oliverqx/sanguine-git-changes-action). Its a clone of sanguine, without the git history. I removed all github workflows and added a very simple workflow that uses the image.

I also opened [4 PRs](https://github.com/oliverqx/sanguine-git-changes-action/pulls) and they all return modules modified based on package dependencies.

Its a bit difficult to follow as sanguine is quite complex and the dependencies can run deep, so I created this other repository, that is easier to follow as its a smaller example. PRs also explain package dependencies quite well, starting here: https://github.com/oliverqx/showcase/pull/1

The test should also pass.

**Comment by trajan0x** [comment]
@oliverqx would you mind running go `generate ./...`?

Also going to merge this in #2391 just to make sure nothing breaks, then can gtm

**Comment by oliverqx** [comment]
Just ran go generate, looks like the solution will be accepted so I'll update documentation.

**Comment by trajan0x** [comment]
Note to self: would like to investigate why 89d9aa9e2ca3ada2f10e89fb69d2e393e6624b6d triggered `contrib/promexporter` as I don't see how that dependency would've been triggered

**Comment by oliverqx** [comment]
> Note to self: would like to investigate why [89d9aa9](https://github.com/synapsecns/sanguine/commit/89d9aa9e2ca3ada2f10e89fb69d2e393e6624b6d) triggered `contrib/promexporter` as I don't see how that dependency would've been triggered

Looking into this, quite a strange bug

**Comment by oliverqx** [comment]
Would you let me know if the README.md is accurate and understandable

**Comment by oliverqx** [comment]
@trajan0x today im pretty much free, will be looking into that bug. I think Its got to do with ct.HasPath() might need to do some edits there, will keep you updated

**Comment by trajan0x** [comment]
Appreciate it

**Comment by oliverqx** [comment]
@trajan0x Actually I realized its not a bug. 

I created this PR in my own sanguine replica: https://github.com/oliverqx/sanguine-git-changes/pull/4

And I added an fmt.Println to all ./contrib/promexporter dependencies to my [image](https://github.com/users/oliverqx/packages/container/package/git-changes-action) and it looks like ./ethergo/submitter is one of the dependencies. Keep in mind that this array a contains all transitive dependencies, so manually following it might be quite complicated. 

Screenshot 2024-04-02 at 16 44 41

**Comment by oliverqx** [comment]
@trajan0x just checking in, I think the solution is ready. Please let me know is there's anything else I could do

**Comment by trajan0x** [comment]
Okay, will stage the latest build today

**Comment by trajan0x** [comment]
LGTM

**Comment by oliverqx** [comment]
sounds good, thanks!

**Comment by oliverqx** [comment]
@trajan0x once the bounty has been settled, I can provide support if anything is wrong. I know my solution is right, but nonetheless. 

Was also wondering if I could get a linkedin review or a recommendation for my personal blog. Would greatly appreciate your input

**Comment by trajan0x** [comment]
Absolutely. Wanna dm me somewhere?

**Comment by oliverqx** [comment]
can i get your discord?

**Comment by oliverqx** [comment]
@trajan0x mine is @olxverqx, or oliver@quixchan.com

**Pull request merged**